---
title: "Introduction to single-cell RNA-seq analysis - Data sets and analyses"
author: "Stephane Ballereau"
date: "Nov 2021"
output:
  ioslides_presentation:
    logo: ../Images/uniOfCamCrukLogos.png
    smaller: yes
    widescreen: yes
    css: css/stylesheet.css
  slidy_presentation: default
  beamer_presentation: default
---
    
## Data sets

* Childhood acute lymphoblastic leukemia (cALL)
  * [Caron et al. 2020](https://www.nature.com/articles/s41598-020-64929-x)
  * the most common pediatric cancer
  * characterized by bone marrow lymphoid precursors
  * that acquire genetic alterations,
  * resulting in disrupted maturation and uncontrollable proliferation 
  * up to 85â€“90% of patients are cured
  * others do not respond to treatment or relapse and die
  * **Aim**: characterise the heterogeneity of gene expression at the cell level, within and between patients
  * **cells**: Bone Marrow Mononuclear cells (BMMCs)

* Adult bone marrow

  * ['HCA': adult BMMCs](https://data.humancellatlas.org/explore/projects/cc95ff89-2e68-4a08-a234-480eca21ce79) (ABMMCs) obtained from the Human Cell Atlas (HCA)

## Samples

Four types of sample are considered:

<!--
* eight patients:
  * six B-ALL
    * four 't(12;21)' or 'ETV6-RUNX1'
    * two 'High hyper diploid' or 'HHD'
  * two T-ALL ('PRE-T')
* three healthy pediatric controls
* eight healthy adult controls, publicly available
-->

* B-ALL patients:
  * four 't(12;21)' or **'ETV6-RUNX1'**
  * two 'High hyper diploid' or **'HHD'**
* T-ALL patients
  * two **'PRE-T'**
* three **healthy pediatric controls**
* eight **healthy adult controls**, publicly available

As the study aims at identifying cell populations, large numbers of cells were sequenced with the droplet-based 10x Chromium assay.

## Analyses

We will follow several steps:

* sequencing quality check
* alignment of reads to the human genome (GRCh38) with 10x software cellranger
* quality control (cell calls, cells and genes filtering)
* UMI count normalisation
* feature selection and dimensionality reduction
* data set integration (PBMMC and ETV6-RUNX1)
* clustering
* identification of cluster marker genes
* differential expression and abundance between conditions
* (trajectory analysis)

## Workflow

```{r echo=FALSE, out.width='60%', fig.align='center'}
knitr::include_graphics('Images/workflow2.png')
```

