---
title: "CRUK CI Summer School 2020 - introduction to single-cell RNA-seq analysis"
subtitle: 'Sequence Quality'

author: "Stephane Ballereau & Ashley Sawle"
output:
  html_notebook:
    code_folding: hide
    toc: yes
    toc_float: yes
    number_sections: true
  html_document:
    df_print: paged
    toc: yes
    number_sections: true
    code_folding: hide
---

```{r setup, include=FALSE}
library(DT)
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE, echo=FALSE)
knitr::opts_chunk$set(cache=FALSE)
```


# Introduction

We will use two sets of Bone Marrow Mononuclear Cells (BMMC):

* 'CaronBourque2020': pediatric samples
* 'HCA': HCA Census of Immune Cells for adult BMMCs

Fastq files were retrieved from publicly available archive (SRA and HCA). 

The samples included are:

```{r sampleTab}
readr::read_tsv("Data/sample_sheet.tsv") %>%  
	as.data.frame() %>%
	datatable(rownames = FALSE, options = list(dom="t", pageLength=20))
```

# 10x library file structure (version 3 chemistry)

## Sequences

The 10x library contains four pieces of information, in the form of DNA sequences,
for each "fragmenti/insert".

* The **sample index** identifies the library, with one or two indexes per sample
* The 10X **cell barcode** (or cell index) identifies the droplet in the library
* The **UMI** identifies the transcript molecule within a cell and gene
* The **insert** is the transcript molecule, ie the cDNA sequence

Previously, only single indexing was used for the samples, more recently 10x
have introduced dual-indexing. This is primarily useful in mitigating the 
effects of index swapping, which can occur when samples are mixed. Though a very
rare event, if our samples are only indexed by a single index, then an index
swap will make a read appear to come from the wrong sample. Dual indexing
avoids this as it would require both indexes to be swapped for the both of the
indexes from another sample, an extremely unlikely event.

The different sample barcodes are identified by a code; "SIGA[A-H][1-12]" for 
single indexing and "SITT[A-H][1-12]" for dual indexing. Specifications for the
single indexes can be found
[here](https://support.10xgenomics.com/single-cell-gene-expression/index/doc/specifications-sample-index-sets-for-single-cell-3)
and for the dual indexes
[here](https://support.10xgenomics.com/spatial-gene-expression/sequencing/doc/specifications-sample-index-sets-for-visium-spatial-gene-expression)

For single sample indexes, reads from each sample are labelled with any one of
a set of four possible 8-nt oligo (8 nt paired oligos if dual indexed) .  For
example, sample index SIGAB8 is 'AAAGTGCT-GCTACCTG-TGCTGTAA-CTGCAAGC'.  All
four are used and are identified by a digit, eg 1-4.  In this way, the number
of possible unique SampleIndex+UMI combinations is increased four-fold.
Depending on the processing pipeline, fastq files may be returned for each
8-ntd index, or combined into a single file.

![](Images/tenxLibStructureV3.png)

For dual indexs, each index has just one I7 and one I5 index, but the I5 varies
with the library prep protocol. e.g.  SITTB6 is I7 = AATGCCATGA and I5 =
TACGTAATGC (for forward stranded workflow) or I5 = GCATTACGTA (for reverse
stranded workflow).

![](Images/tenxLibStructureV3dual.png)

## Raw fastq files

The sequences for any given fragment will generally be delivered in 3 or 4
files:

* **I1**: I7 sample index
* **I2**: I5 sample index if present (dual indexing only)
* **R1**: 10x barcode + UMI
* **R2**: insert sequence

Depending on the exact sequencing processing pipeline, instances of multiple
sample indexes for single indexed libraries may be combined into a single file
or maybe be provided separately (meaning 12 fastq files per sample, per lane of
sequencing).  

Similarly, naming conventions for the files will vary according to the processing
pipeline, but at the least the filenames need to include the sample barcode ID. 
The following is an example of the file naming convention for files delivered 
from the CRUK Genomics Core 10x processing pipeline:








`r knitr::knit_exit()`




## CaronBourque2020 - fastqc

```{r}
# CaronBourque2020
cb_sampleSheetFn <- file.path(projDir, "Data/CaronBourque2020/SraRunTable.txt")
cb_sampleSheet <- read.table(cb_sampleSheetFn, header=T, sep=",")
#cb_sampleSheet <-  cb_sampleSheet %>% filter(!Run == "SRR9264351")
cb_sampleSheet
```

```{r, results = 'asis'}
htmlVec <- list.files(fastqcDir)
htmlVec <- grep("\\.html$", htmlVec, value=TRUE)
```

```{r}
filesDf <- data.frame(
		      "I1" = sprintf("%s_S0_L001_%s_001_fastqc.html", cb_sampleSheet$Run, "I1"),
		      "R1" = sprintf("%s_S0_L001_%s_001_fastqc.html", cb_sampleSheet$Run, "R1"),
		      "R2" = sprintf("%s_S0_L001_%s_001_fastqc.html", cb_sampleSheet$Run, "R2")
)
rownames(filesDf) <- cb_sampleSheet$Run
```

```{r, results = 'asis'}
for (runx in cb_sampleSheet$Run)
{
	cat("Run ", runx, ":\n\n")
	for(i in c("I1", "R1", "R2"))
	{
		#filepath <- file.path(fastqcDir, filesDf[runx,i])
		filepath <- file.path(fastqcDirLink, filesDf[runx,i])
		cat(i, ": [", filesDf[runx,i], "](",filepath,")\n\n")
	}
}
```

## CaronBourque2020 - MultiQC

### sample index: I1

```{r, results = 'asis'}
htmlVec <- list.files(paste0(fastqcDir, "/Multiqc/I1"))
htmlVec <- grep("\\.html$", htmlVec, value=TRUE)
for(i in htmlVec){
	filename <- file.path(fastqcDirLink, "/Multiqc/I1", i)
	cat("[", i, "](",filename,")\n\n")
}
```

### cell barcode + UMI: R1

```{r, results = 'asis'}
htmlVec <- list.files(paste0(fastqcDir, "/Multiqc/R1"))
htmlVec <- grep("\\.html$", htmlVec, value=TRUE)
for(i in htmlVec){
	filename <- file.path(fastqcDirLink, "/Multiqc/R1", i)
	cat("[", i, "](",filename,")\n\n")
}
```

### insert: R2

```{r, results = 'asis'}
htmlVec <- list.files(paste0(fastqcDir, "/Multiqc/R2"))
htmlVec <- grep("\\.html$", htmlVec, value=TRUE)
for(i in htmlVec){
	filename <- file.path(fastqcDirLink, "/Multiqc/R2", i)
	cat("[", i, "](",filename,")\n\n")
}
```

## HCA adult BMMC - fastqc

For the HCA adult BMMC fastq files were provided for each 8-ntd sample index and lane. We ran fastqc on each separately. We are therefore not listing links to the fastqc reports but only to the MultiQC reports.

```{r}
fastqcDir <- sprintf("%s/Data/%s/fastqc", projDir, "Hca")
fastqcDirLink <- sprintf("%s/Data/%s/fastqc", projDirLink, "Hca")

# HCA
hca_sampleSheetFn <- file.path(projDir, "Data/Hca/accList_Hca.txt")

hca_sampleSheet <- read.table(hca_sampleSheetFn, header=F, sep=",")
colnames(hca_sampleSheet) <- "Run"
hca_sampleSheet
```

```{r, results = 'asis'}
htmlVec <- list.files(fastqcDir)
htmlVec <- grep("\\.html$", htmlVec, value=TRUE)
```

`r #length(htmlVec)`
378 fastqc reports were compiled in the multiQC reports below.

##  HCA adult BMMC - MultiQC

### sample index: I1

```{r, results = 'asis'}
htmlVec <- list.files(paste0(fastqcDir, "/Multiqc/I1"))
htmlVec <- grep("\\.html$", htmlVec, value=TRUE)
for(i in htmlVec){
	filename <- file.path(fastqcDirLink, "/Multiqc/I1", i)
	cat("[", i, "](",filename,")\n\n")
}
```

### cell barcode + UMI: R1

```{r, results = 'asis'}
htmlVec <- list.files(paste0(fastqcDir, "/Multiqc/R1"))
htmlVec <- grep("\\.html$", htmlVec, value=TRUE)
for(i in htmlVec){
	filename <- file.path(fastqcDirLink, "/Multiqc/R1", i)
	cat("[", i, "](",filename,")\n\n")
}
```

### insert: R2

```{r, results = 'asis'}
htmlVec <- list.files(paste0(fastqcDir, "/Multiqc/R2"))
htmlVec <- grep("\\.html$", htmlVec, value=TRUE)
for(i in htmlVec){
	filename <- file.path(fastqcDirLink, "/Multiqc/R2", i)
	cat("[", i, "](",filename,")\n\n")
}
```


