---
title: "Introduction to single-cell RNA-seq analysis"
subtitle: 'Quality Control'
author: "Stephane Ballereau, Ashley Sawle"
output:
  html_document:
    df_print: paged
    toc: yes
    toc_depth: 2
    number_sections: true
    code_folding: show 
    css: ../css/boxes.css
---

```{r setup, echo=FALSE, include=FALSE, message=FALSE}
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE, cache=TRUE)
knitr::opts_chunk$set(fig.width=7, fig.height=7)
knitr::opts_chunk$set(dev="CairoPNG")
set.seed(123)
```

# Introduction

In the course materials we performed QC and filtering of 2 samples from each
of the sample groups. For this challenge we would like you to perform QC and 
filtering on all of the samples from the Caron data set. 

# Load packages

```{r libraries}
library(DropletUtils)
library(scater)
library(ensembldb)
library(AnnotationHub)
library(BiocParallel)
library(tidyverse)
library(patchwork)
library(ggvenn)
```

# Sample meta data

We will load both the Caron and HCA data sets. We have already prepared a sample
meta data table that relates the sample/run ID to the sample group.

```{r samplesheet, message=FALSE}
samplesheet <- read_tsv("Data/sample_sheet.tsv")
```

**NOTE**: There are two PBMMC_1 samples. These are two libraries from the same
sample material. Later on, we will combine these to form a single sample, but 
for now we will need to rename them so that they are QC'd separately.

# Load data

## Parallelisation

We first need to first set up some parallel parameters using the package
`BiocParallel`.

```{r parallel}
bp.params <- MulticoreParam(workers = 7)
```

:::exercise 
In order to load the CellRanger data for all of the Caron samples, you will first
need to create a named vector of the paths to the filtered count matrix folders 
called `list_of_files` and then use this in the `read10xCounts` command. 

<details><summary>Hint</summary>
The paths to `filtered_feature_bc_matrix` directories for each sample can be
constructed using the **SampleId** as:   

<span style = "color: #545454; font-weight: bold;">CellRanger_Outputs/<span style="color: #2e2892; font-style: italic;">SampleId</span>/outs/filtered_feature_bc_matrix</span>

You will need to use a function such as `str_c` or `paste`

The names of the vector will determine the sample name used in the counts 
matrix, this should be the 'SRR' number as in the SampleId column of the
sample sheet.
</details>


<details><summary>Answer</summary>
```{r loadData}
samples_list <- samplesheet %>% 
    filter(DatasetName=="Caron") %>%  
    pull(SampleId)
list_of_files <- str_c("CellRanger_Outputs/", 
                       samples_list, 
                       "/outs/filtered_feature_bc_matrix")
names(list_of_files) <- samples_list

    
sce <- read10xCounts(list_of_files, col.names=TRUE, BPPARAM = bp.params)
```
</details>

:::

## Check samples in the data set

A good sanity check is to look at the colData to ensure that we have all the
samples we are expecting and check the overall size of the new object.

```{r checkDataObject}
colData(sce) %>%
    as.data.frame() %>% 
    select(Sample) %>% 
    distinct()
sce
```

## Modify the droplet annotation

We need to modify the cell barcodes so that they unique for each sample. We should
also modify the names of the two PBMMC_1 samples, so that we can distinguish them
from one another.

We will also switch the "Sample" column to be the sample name and add 
information from the sample sheet to the droplet annotation.

:::exercise

In the code below, replace the "**XXXXXXXX**" to rename the PBMMC_1 samples
appropriately.

<pre class="r"><code>colData(sce) &lt;- colData(sce) %&gt;% 
    as.data.frame() %&gt;%
    rownames_to_column(&quot;RowName&quot;) %&gt;% 
    mutate(SampleNum = str_extract(RowName, &quot;^[0-9]+&quot;)) %&gt;%
    mutate(Barcode = str_replace(Barcode, &quot;1$&quot;, SampleNum)) %&gt;%
    left_join(samplesheet, by=c(Sample=&quot;SampleId&quot;)) %&gt;%
    rename(SampleId=Sample) %&gt;% 
    rename(Sample=SampleName) %&gt;%    
    mutate(Sample = case_when(
           SampleId == <span style = "font-weight: bold;">XXXXXXXX</span> ~ str_c(Sample, &quot;a&quot;),
           SampleId == <span style = "font-weight: bold;">XXXXXXXX</span> ~ str_c(Sample, &quot;b&quot;),
           TRUE ~ Sample)) %&gt;% 
    column_to_rownames(&quot;RowName&quot;) %&gt;% 
    select(Sample, Barcode, SampleId, SampleGroup, DatasetName) %&gt;%
    DataFrame()</code></pre>

<details><summary>Hint</summary>
The new `SampleId` columns contains the "SRR" numbers, as does the SampleId in
the sample sheet. Replace the **XXXXXXXXX** with the appropriate "SRR" number
by checking for the PBMMC_1 samples in the samplesheet.
</details>

<details><summary>Answer</summary>
```{r dataSets_addSampleSheetAnswer}
colData(sce) <- colData(sce) %>% 
    as.data.frame() %>%
    rownames_to_column("RowName") %>% 
    mutate(SampleNum = str_extract(RowName, "^[0-9]+")) %>%
    mutate(Barcode = str_replace(Barcode, "1$", SampleNum)) %>%
    left_join(samplesheet, by=c(Sample="SampleId")) %>%
    rename(SampleId=Sample) %>% 
    rename(Sample=SampleName) %>%    
    mutate(Sample = case_when(
           SampleId == "SRR9264351" ~ str_c(Sample, "a"),
           SampleId == "SRR9264352" ~ str_c(Sample, "b"),
           TRUE ~ Sample)) %>% 
    column_to_rownames("RowName") %>% 
    select(Sample, Barcode, SampleId, SampleGroup, DatasetName) %>%
    DataFrame()
```
</details>

:::

Before moving on, check that the columns are as we expect them to be and that
the PBMMC_1 samples have been renamed:

```{r checkColData}
colData(sce)
unique(sce$Sample)
```

# Undetected genes

Remove undetected genes.

```{r detected_genes}
detected_genes <- rowSums(counts(sce)) > 0
sce <- sce[detected_genes,]
```

:::exercise

What proportion of genes have been detected

<details><summary>Answer</summary>
```{r percentageDetected}
detected <- sum(detected_genes)/length(detected_genes)
detected
```

Approximately `r round(detected, 3)*100` percent of genes were detected in at
least one sample.
</details>

:::

# Annotate genes

In order to assess the percentage of mitochondrial UMIs, we will need to be
able to identify mitochondrial genes. The simplest way to do this is to annotate
the genes with their chromosome of origin.

```{r annotate_genes}
ah <- AnnotationHub()
ens.mm.98 <- query(ah, c("Homo sapiens", "EnsDb", 98))[[1]] 

genes <- rowData(sce)$ID
gene_annot <- AnnotationDbi::select(ens.mm.98, 
                                    keys = genes,
                                    keytype = "GENEID",
                                    columns = c("GENEID", "SEQNAME")) %>%
    set_names(c("ID", "Chromosome"))
rowData(sce) <- merge(rowData(sce), gene_annot, by = "ID", sort=FALSE)
rownames(rowData(sce)) <- rowData(sce)$ID

rowData(sce)
```


# Add per cell QC metrics

Now add the per cell QC metrics to the droplet annotation using the
function `addPerCellQC`.

```{r qc_addPerCellQC}
is.mito <- which(rowData(sce)$Chromosome=="MT")

sce <- addPerCellQC(sce, subsets=list(Mito=is.mito), BPPARAM = bp.params)
```

# Explore QC metric distribution

Before moving on to do the actual cell filtering, it is always a good idea to
explore the distribution of the metrics across the droplets.

:::exercise
Use the `scater` function `plotColData` to generate plots showing the
distributions of the total number of UMIs, the number of genes detected and
percentage of UMIs aligned to mitochondrial genes across all cells for each
sample.

<details><summary>Answer</summary>
```{r eval = FALSE, fig.width=12, fig.height=4}
plotColData(sce, x="Sample", y="sum",other_fields="SampleGroup") + 
    facet_wrap(~SampleGroup, nrow=1, scales = "free_x") + 
    scale_y_log10() + 
    ggtitle("Total count")
```

```{r eval = FALSE, fig.width=12, fig.height=4}
plotColData(sce, x="Sample", y="detected", other_fields="SampleGroup") + 
    facet_wrap(~SampleGroup, nrow=1, scales = "free_x") + 
    scale_y_log10() + 
    ggtitle("Detected features")
```

```{r eval = FALSE, fig.width=12, fig.height=4}
plotColData(sce, x="Sample", y="subsets_Mito_percent", other_fields="SampleGroup") + 
    facet_wrap(~SampleGroup, nrow=1, scales = "free_x") +
    ggtitle("Mito percent")
```
</details>

:::

```{r umiDistr, echo = FALSE, fig.width=12, fig.height=4}
plotColData(sce, x="Sample", y="sum",other_fields="SampleGroup") + 
    facet_wrap(~SampleGroup, nrow=1, scales = "free_x") + 
    scale_y_log10() + 
    ggtitle("Total count")
```

```{r geneDistr, echo = FALSE, fig.width=12, fig.height=4}
plotColData(sce, x="Sample", y="detected", other_fields="SampleGroup") + 
    facet_wrap(~SampleGroup, nrow=1, scales = "free_x") + 
    scale_y_log10() + 
    ggtitle("Detected features")
```

```{r mtDistr, echo = FALSE, fig.width=12, fig.height=4}
plotColData(sce, x="Sample", y="subsets_Mito_percent", other_fields="SampleGroup") + 
    facet_wrap(~SampleGroup, nrow=1, scales = "free_x") +
    ggtitle("Mito percent")
```

# Identification of low-quality cells with adaptive thresholds

:::exercise

Use the scater function `quickPerCellQC` to assess cell quality based
on the three metrics. Name the object generated `cell_qc_results`.  

When running the command, consider the distribution plots above and decide
whether to use the `batch` option and if so, at what level it should be
applied.

How many cells will be removed from the data set?


<details><summary>Hint</summary>
With these samples we have three possible "batch" levels at which we could 
run adaptive filtering. We could apply the filtering across all samples 
together (i.e. no `batch`), we could apply it by sample group (ETV6-RUNX1, 
HHD, PBMMC, PRE-T), or we could apply it per sample. 
</details>


<details><summary>Answer</summary>
It is clear that there are significantly different distribution profiles
between samples in the same group. Noticeably the distribution profiles for the
two PBMMC_1 samples are quite different - it would appear that PBMMC_1b has a
lower overal UMI count indicating less sequencing. For this reason it may be
prudent to apply the adaptive filtering to each sample independently.

```{r quickPerCellQC_batch_compute}
cell_qc_results <- quickPerCellQC(colData(sce),
                                percent_subsets=c("subsets_Mito_percent"),
                                batch=sce$Sample)
as.data.frame(cell_qc_results) %>% summarise(across(everything(), sum))
```

In total `r sum(cell_qc_results$discad)` cells will be removed from the
dataset.

</details>

:::

Let's replace the columns in the droplet annotation with these new filters.

```{r quickPerCellQC_batch_replace}
sce$low_lib_size <- cell_qc_results$low_lib_size
sce$low_n_features <- cell_qc_results$low_n_features
sce$high_Mito_percent <- cell_qc_results$high_subsets_Mito_percent
sce$discard <- cell_qc_results$discard
```

We can visualise how the new filters look using violin plots.

```{r filtUMI, fig.width=12, fig.height=4}
plotColData(sce, 
            x="Sample", 
            y="sum",
            other_fields="SampleGroup", 
            colour_by = "low_lib_size") + 
    facet_wrap(vars(SampleGroup), nrow=1, scales = "free_x") + 
    scale_y_log10() + 
    labs(y = "Total count", title = "Total count") +
    guides(colour=guide_legend(title="Discarded"))
```

```{r filtGene, fig.width=12, fig.height=4}
plotColData(sce, 
            x="Sample", 
            y="detected",
            other_fields="SampleGroup", 
            colour_by = "low_n_features") + 
    facet_wrap(vars(SampleGroup), nrow=1, scales = "free_x") + 
    scale_y_log10() + 
    labs(y = "Genes detected", title = "Genes detected") +
    guides(colour=guide_legend(title="Discarded"))
```

```{r filtMT, fig.width=12, fig.height=4}
plotColData(sce, 
        x="Sample", 
        y="subsets_Mito_percent",
        other_fields="SampleGroup", 
        colour_by = "high_Mito_percent") + 
    facet_wrap(vars(SampleGroup), nrow=1, scales = "free_x") + 
    labs(y = "Percentage mitochondrial UMIs",
         title = "Mitochondrial UMIs") +
    guides(colour=guide_legend(title="Discarded"))
```

# Filtering out poor quality cells

Filter out the poor quality cells, recalculate the cell QC metrics and save the
filtered object.

```{r eval = FALSE}
sce <- sce[, !sce$discard]
colData(sce) <- colData(sce)[,1:3]
sce <- addPerCellQC(sce, BPPARAM = bp.params)
saveRDS(sce, "../results/Caron_filtered_genes.rds")
```

# Filtering genes based on sparsity - Bonus exercise

:::exercise 

So far we have only eliminated genes that are undetected across all samples. In reality
there will be a large number of genes that are only detected at low levels in a small
number of cells - much of this will just be due to technical noise. We could further 
filter the genes using "sparsity". 

Look at section 8 of the documention. Filter the dataset to remove genes that 
have only been detected in fewer than 20 cells.

How many additional genes are filtered out?

<details><summary>Answer</summary>
```{r qc_addPerFeatureQC, eval=TRUE}
sce <- addPerFeatureQC(sce, BPPARAM = bp.params)
rowData(sce)$gene_sparsity <- (100 - rowData(sce)$detected) / 100
max.sparsity <- 1 - (20 / ncol(sce))
sparse.genes <- rowData(sce)$gene_sparsity > max.sparsity
sce <- sce[!sparse.genes, ]
sum(sparse.genes)
```

An additional `r sum(sparse.genes)` have been filtered out.

</details>

:::

# Session information

<details>
```{r}
sessionInfo()
```
</details>
