---
title: "Introduction to single-cell RNA-seq analysis"
subtitle: 'Normalisation - GSM3872434 sample'

author: "Stephane Ballereau"
output:
  html_document:
    df_print: paged
    toc: yes
    number_sections: true
    code_folding: show
  html_notebook:
    code_folding: hide
    toc: yes
    toc_float: yes
    number_sections: true
  html_book:
    code_folding: hide
---

# Normalisation - Exercises

Exercise: apply the deconvolution and SCTransform normalisations methods on a
single sample: ETV6-RUNX1_1 (aka GSM3872434).

```{r norm_Caron.knitr_options, echo=FALSE, results="hide", message=FALSE}
require(knitr)
# set working directory# set working directory# set working directory
knitr::opts_knit$set(root.dir = '/home/ubuntu/Course_Materials/Exercises')
# set other chunk variables
opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE, eval=FALSE)
opts_chunk$set(fig.width=7, fig.height=7)
set.seed(123) # for reproducibility
```

```{r libraries_norm_GSM3872434_allCells, include=FALSE, results='hide', message=FALSE, warning=FALSE}
library(scater)
library(scran)
library(tidyverse)
library(BiocSingular)
library(BiocParallel)

bpp <- MulticoreParam(7)
```

## Load object

```{r GSM3872434_readIn}
# Read object in:
# remember getwd() and dir()
sce <- readRDS("../Robjects/Caron_filtered.rds")
colData(sce)$SampleName <- colData(sce)$Sample
```

Select cells for ETV6-RUNX1_1:

```{r GSM3872434_downsample}
# have new list of cell barcodes for each sample
sce.master <- sce
vec.bc <- colData(sce.master) %>%
	data.frame() %>%
	filter(SampleName == "ETV6-RUNX1_1") %>%
	group_by(SampleName) %>%
	pull(Barcode)
```

Number of cells in the sample:

```{r GSM3872434_barcodes}
table(colData(sce.master)$Barcode %in% vec.bc)
```

Subset cells from the SCE object:

```{r GSM3872434_subset_sce}
tmpInd <- which(colData(sce.master)$Barcode %in% vec.bc)
sce <- sce.master[,tmpInd]
sce
```

Check columns data:

```{r GSM3872434_SampleName_table}
head(colData(sce))
table(colData(sce)$SampleName)
```

<!--
# mind that genes were filtered using all cells, not just those sampled here.
-->

## Exercise 1 : Deconvolution

Cluster cells then normalise.

### Cluster cells

```{r comp_quickClus_norm_GSM3872434_allCells}
set.seed(100) # clusters with PCA from irlba with approximation
clust <- quickCluster(sce) # slow with all cells.
table(clust)
```

### Compute size factors

```{r calculateSumFactors_norm_GSM3872434_allCells}
# deconvolve


# seet size factors


# size factors distribution summary


```

Plot deconvolution size factors against library size factors:

```{r scatter_deconvSf_libSf_prep_norm_GSM3872434_allCells}
# compute library size factors


# make data frame keeping library and deconvolution size factors for plotting


```

```{r scatter_deconvSf_libSf_colBy_more_norm_GSM3872434_allCells}
# plot deconv.sf against lib.sf

# colour by library size


```

### Apply size factors

<!--
For each cell, raw counts for genes are divided by the size factor for that cell
and log-transformed so downstream analyses focus on genes with strong relative
differences. We use `scater::logNormCounts()`.
-->

```{r logNormCounts_norm_GSM3872434_allCells}


```

## Exercise 2 : sctransform

<!--
With scaling normalisation a correlation remains between the mean and variation
of expression (heteroskedasticity). This affects downstream dimensionality
reduction as the few main new dimensions are usually correlated with library
size. SCTransform addresses the issue by regressing library size out of raw
counts and providing residuals to use as normalized and variance-stabilized
expression values in downstream analysis. We will use the
[sctransform vignette](https://cran.r-project.org/web/packages/sctransform/index.html).
-->

Get UMI counts matrix:

```{r counts_sct_GSM3872434_allCells}
counts <- counts(sce)
colnames(counts) <- colData(sce)$Barcode
```

### Inspect data

<!--
We will now calculate some properties and visually inspect the data. Our main
interest is in the general trends not in individual outliers. Neither genes nor
cells that stand out are important at this step, but we focus on the global trends.
-->

Derive gene and cell attributes from the UMI matrix.

```{r attr_comp_sct_GSM3872434_allCells}






```

Mean-variance relationship plot

<!--
For the genes, we can see that up to a mean UMI count of 0 the variance follows
the line through the origin with slop one, i.e. variance and mean are roughly
equal as expected under a Poisson model. However, genes with a higher average
UMI count show overdispersion compared to Poisson.
-->

```{r attr_plot_sct_GSM3872434_allCells}




```

Mean-detection-rate relationship plot

<!--
In line with the previous plot, we see a lower than expected detection rate in the medium expression range. However, for the highly expressed genes, the rate is at or very close to 1.0 suggesting that there is no zero-inflation in the counts for those genes and that zero-inflation is a result of overdispersion, rather than an independent systematic bias.
-->

```{r scatter_detecRate_logMean_sct_GSM3872434_allCells}
# add the expected detection rate under Poisson model



```

### Transformation

<!--
"Based on the observations above, which are not unique to this particular data
set, we propose to model the expression of each gene as a negative binomial
random variable with a mean that depends on other variables. Here the other
variables can be used to model the differences in sequencing depth between
cells and are used as independent variables in a regression model. In order to 
avoid overfitting, we will first fit model parameters per gene, and then use
the relationship between gene mean and parameter values to fit parameters,
thereby combining information across genes. Given the fitted model parameters,
we transform each observed UMI count into a Pearson residual which can be
interpreted as the number of standard deviations an observed count was away
from the expected mean. If the model accurately describes the mean-variance
relationship and the dependency of mean and latent factors, then the result
should have mean zero and a stable variance across the range of expression."
[sctransform vignette](https://cran.r-project.org/web/packages/sctransform/index.html).
-->

Estimate model parameters and transform data

<!--
The vst function estimates model parameters and performs the variance stabilizing
transformation. Here we use the log10 of the total UMI counts of a cell as variable
for sequencing depth for each cell. After data transformation we plot the model
parameters as a function of gene mean (geometric mean).
-->

```{r comp_sct_GSM3872434_allCells, warning=FALSE}
print(dim(counts))
# We use the Future API for parallel processing; set parameters here
future::plan(strategy = 'multicore', workers = 4)
options(future.globals.maxSize = 10 * 1024 ^ 3)

set.seed(44)

# now transform:








```

Inspect model:

```{r model_show_sct_GSM3872434_allCells}
print(vst_out$model_str)
```

We will look at several genes in more detail.

```{r plot_model_1_sct_GSM3872434_allCells}
rowData(sce) %>%
  as.data.frame %>%
  filter(Symbol %in% c('RPL10', 'FTL'))

# plot model data for these two genes:




```

Distribution of residual mean:

```{r plot_model_resMean_sct_GSM3872434_allCells}
ggplot(vst_out$gene_attr, aes(residual_mean)) +
  geom_histogram(binwidth=0.01)
```

Distribution of residual variance:

```{r plot_model_resVar_sct_GSM3872434_allCells}
ggplot(vst_out$gene_attr, aes(residual_variance)) +
  geom_histogram(binwidth=0.1) +
  geom_vline(xintercept=1, color='red') +
  xlim(0, 10)
```

Variance against mean (genes):

```{r plot_model_resVar_gMean_sct_GSM3872434_allCells}
ggplot(vst_out$gene_attr,
       aes(log10(gmean), residual_variance)) +
       geom_point(alpha=0.3, shape=16) +
       geom_density_2d(size = 0.3)
```

Check genes with large residual variance:

```{r table_show_sct_GSM3872434_allCells}
dd <- vst_out$gene_attr %>%
	arrange(-residual_variance) %>%
	slice_head(n = 22) %>%
	mutate(across(where(is.numeric), round, 2))

dd %>% tibble::rownames_to_column("ID") %>%
	left_join(as.data.frame(rowData(sce))[,c("ID", "Symbol")], "ID") %>%
	DT::datatable(rownames = FALSE)
```

Check transformed values:

```{r check_trans_sct_GSM3872434_allCells}
print(dim(vst_out$y))
vst_out$y[1:10,1:5]
```

Genes that are expressed in fewer than 5 cells are not used and not returned,
so to add vst_out$y as an assay we need to remove the missing genes.

```{r subsetGenes_sct_GSM3872434}
sceOrig <- sce
sceOrig
geneOverlap <- rownames(sce) %in% rownames(vst_out$y)
if(!all(geneOverlap))
{
  table(rownames(sce) %in% rownames(vst_out$y))
  tmpInd <- which(rownames(sce) %in% rownames(vst_out$y))
  sce <- sce[tmpInd,]
  assayNames(sce)
}
rm(sceOrig)
```
