---
title: "CRUK CI Summer School 2020 - introduction to single-cell RNA-seq analysis"
subtitle: 'batch correction - 500 cells per sample'

output:
  html_document:
    df_print: paged
    toc: yes
    number_sections: true
    code_folding: hide
  html_book:
    code_folding: hide
---

# Batch correction - GSM3872442 set

When we do single cell RNA-Seq experiments most of the time inevitably there will be technical batch effects. For example, logistically maybe we can't put all of our samples in the same run and so have to split them up. Even within samples all sequenced together there can be batch effects from the previous steps. We need to check and correct for them.

## Learning objectives

Understand different batch correction methods using one sample sequenced in two different runs. We will look at differences in :

* Normalisation order

* Correction with limma

* Correction with mnnCorrect

* Correction with fastMNN

* Correction with Harmony

## Data

GSM3872442 is a single PBMMC sample sequenced as a pool of two libraries:
SRR9264351 and SRR9264352.

We will use this sample to illustrate batch correction methods.

```{r libraries_batch_GSM3872442, message=FALSE, include=FALSE}
library(scater)
library(scran)
library(ggplot2)
library(dplyr)
library(BiocSingular)
library(Cairo)
```

Load object 

```{r GSM3872442_readIn}
sce <- readRDS("~/Downloads/caron_sce_nz_postQc.Rds")
```

Select the GSM3872442 cells:

```{r subset_cells_batch_GSM3872442, eval=FALSE}
sample1.nz.sce <- SingleCellExperiment(list(counts=counts(sce[, sce$Run %in% c("SRR9264351")])),
                                       colData=colData(sce[, sce$Run %in% c("SRR9264351")]))

sample2.nz.sce <- SingleCellExperiment(list(counts=counts(sce[, sce$Run %in% c("SRR9264352")])),
                                       colData=colData(sce[, sce$Run %in% c("SRR9264352")]))

saveRDS(sample1.nz.sce, "Robjects/BC_sample1.Rds")
saveRDS(sample2.nz.sce, "Robjects/BC_sample2.Rds")
```
The above two code chunks just illustrate how the starting demonstration files for this chapter were made, if you are working from our github repo you will not need this. if you are testing batch corrections on your own samples, for this first part, read in each sample as a separate sce object.

```{r}
sample1.sce <- readRDS("Robjects/BC_sample1.Rds")
sample2.sce <- readRDS("Robjects/BC_sample2.Rds")
```

## Normalise each separately and re-pool

```{r sep_norm_batch_GSM3872442}
sample1.qclusters <- quickCluster(sample1.sce, method="igraph")
sample1.sce <- computeSumFactors(sample1.sce, min.mean=0.1, cluster=sample1.qclusters)
sample1.sce <- logNormCounts(sample1.sce)

sample2.qclusters <- quickCluster(sample2.sce, method="igraph")
sample2.sce <- computeSumFactors(sample2.sce, min.mean=0.1, cluster=sample2.qclusters)
sample2.sce <- logNormCounts(sample2.sce)

```

Re-pool:

```{r repool_batch_GSM3872442}
# recombine the normalized samples together
all.samp.exprs <- do.call(cbind,
                          list("SRR9264351"=exprs(sample1.sce),
                               "SRR9264352"=exprs(sample2.sce)))
colnames(all.samp.exprs) <- c(as.character(colData(sample1.sce)$Barcode),
                              as.character(colData(sample2.sce)$Barcode))
```

For the PCA we want to quickly select the genes that are most informative. We will use the top 2000 genes with the highest variance.

```{r sep_hvg_batch_GSM3872442}
gene.variances <- apply(all.samp.exprs, 1, var)
names(gene.variances) <- rownames(all.samp.exprs)
highly.variable.genes <- names(gene.variances[order(gene.variances, decreasing=TRUE)])[1:2000]
```

Perform PCA:

```{r sep_comp_pca_batch_GSM3872442}
# we need to use a fast approximate algorithm for PCA on large data sets
# this algorithm has a stochastic component,
# so we need to fix the seed number to get the same result each time
set.seed(42)
separate.hvg.pca <- irlba::prcomp_irlba(t(all.samp.exprs[highly.variable.genes, ]), n=5) # we only need a few components
separate.hvg.pcs <- as.data.frame(separate.hvg.pca$x) # extract the principal components
separate.hvg.pcs$Cell <- colnames(all.samp.exprs) # set the sample column as the cell IDs

# combine the PCs with the sample information into a single data frame for plotting
samples.info <- data.frame("Cell"=colnames(all.samp.exprs),
                           "Run"=c(rep("SRR9264351", ncol(sample1.sce)), 
                                   rep("SRR9264352", ncol(sample2.sce))))

# merge the two data frames together
separate.pca.merge <- merge(separate.hvg.pcs, samples.info, by='Cell')

```
You can merge and plot the PCA results.

```{r sep_plot_pca_batch_GSM3872442}
ggplot(separate.pca.merge, aes(x=PC1, y=PC2, fill=Run)) +
  geom_point(shape=21, size=3) +
  theme_minimal()
```

It is easier to bind the two sce objects together into a single sce (remember to insure the colData has the same columns) and the plot whatever you need.

```{r sep_cbind_batch_GSM3872442}
sce.sep <- cbind(sample1.sce, sample2.sce)

sce.sep <- runPCA(sce.sep)

plotPCA(sce.sep, colour_by="Run", size_by = "sum")
```

```{r sep_tsne_batch_GSM3872442}
sce.sep <- runTSNE(sce.sep, dimred="PCA")
plotTSNE(sce.sep, colour_by="Run", size_by = "sum")
```

```{r sep_umap_batch_GSM3872442}
sce.sep <- runUMAP(sce.sep, dimred="PCA")
plotUMAP(sce.sep, colour_by="Run", size_by = "sum")

```

From the above plots you can see that even though the samples should be the same there is a batch effect from being run seperately. Now we can see it we can try to remove it.

## Normalise batches together

```{r tog_load_GSM3872442, eval=FALSE}
sample3.sce <- SingleCellExperiment(list(counts=counts(sce[, sce$Run %in% c("SRR9264351", "SRR9264352")])),
                                       colData=colData(sce[, sce$Run %in% c("SRR9264351", "SRR9264352")]))

saveRDS(sample3.sce, "./Robjects/BC_sample3.Rds")
```
The above chunk illustrates how the combined object was made (`sample 3`) but if you are using our repo you can just load it in and process it with the preceeding steps.

```{r tog_norm_batch_GSM3872442}

sample3.sce <- readRDS("./Robjects/BC_sample3.Rds")

sample3.qclusters <- quickCluster(sample3.sce, method="igraph")
sample3.sce <- computeSumFactors(sample3.sce, min.mean=0.1, cluster=sample3.qclusters)
sample3.sce <- logNormCounts(sample3.sce)

pool.exprs <- exprs(sample3.sce)
colnames(pool.exprs) <- gsub(colData(sample3.sce)$Barcode, pattern="-", replacement=".")

```

Find the 2000 genes with the highest variance:

```{r tog_hvg_batch_GSM3872442}
gene.variances <- apply(pool.exprs, 1, var)
names(gene.variances) <- rownames(pool.exprs)
highly.variable.genes <- names(gene.variances[order(gene.variances, decreasing=TRUE)])[1:2000]

```

Perform PCA:

```{r tog_comp_pca_batch_GSM3872442}
# we need to use a fast approximate algorithm for PCA on large data sets
# this algorithm has a stochastic component, so we need to fix the seed number to get the same result each time
set.seed(42)
combined.hvg.pca <- irlba::prcomp_irlba(t(pool.exprs[highly.variable.genes, ]), n=5) # we only need a few components
combined.hvg.pcs <- as.data.frame(combined.hvg.pca$x) # extract the principal components
combined.hvg.pcs$Cell <- colnames(pool.exprs) # set the sample column as the cell IDs

# combine the PCs with the sample information into a single data frame for plotting
samples.info <- data.frame("Cell"=colnames(pool.exprs),
                           "Run"=colData(sample3.sce)$Run)

# merge the two data frames together
combined.pca.merge <- merge(combined.hvg.pcs, samples.info, by='Cell')


```

Plot PC1-PC2 plane, with cells colored by 'Run' (and sized according to library size):

```{r tog_show_pca_batch_GSM3872442}
sample3.sce <- runPCA(sample3.sce)
plotPCA(sample3.sce, colour_by="Run", size_by = "sum")
```
Plot the TSNE:

```{r tog_show_tsne_batch_GSM3872442}
sample3.sce <- runTSNE(sample3.sce, dimred="PCA")
plotTSNE(sample3.sce, colour_by="Run", size_by = "sum")
```
Plot the UMAP:

```{r tog_show_umap_batch_GSM3872442}
sample3.sce <- runUMAP(sample3.sce, dimred="PCA")
plotUMAP(sample3.sce, colour_by="Run", size_by = "sum")
```

Notice there is a difference o the plots between running these step separately or together. 

## Batch correction

Make sure you sce object contains the information about which run and that information is a factor. This will be needed for future steps.

```{r editSce_batch_GSM3872442}
sample3.sce$Run <- factor(sample3.sce$Run)
sample3.sce$batch <- sample3.sce$Run
```

###  Gaussian (normal) linear models

We can use Limma functions to remove the batch effect using a linear model.

```{r linReg_batch_GSM3872442}
suppressMessages(require(limma))
lm_design_batch <- model.matrix(~0 + batch, data = colData(sample3.sce))
fit_lm_batch <- lmFit(logcounts(sample3.sce), lm_design_batch)
resids_lm_batch <- residuals(fit_lm_batch, logcounts(sample3.sce))
assay(sample3.sce, "lm_batch") <- resids_lm_batch

reducedDim(sample3.sce, "PCA_lm_batch") <- reducedDim(
  runPCA(sample3.sce, exprs_values = "lm_batch"), "PCA")

plotReducedDim(sample3.sce, dimred = "PCA_lm_batch",
        colour_by = "batch", 
        size_by = "sum",
        shape_by = "Sample.Name"
        ) +
  ggtitle("LM - regress out batch")
```

However, as you can see from the PCA the batch effect has not been entirely removed.

## mnnCorrect

mnnCorrect uses a mutual nearest neighbour approach to batch correct. There is extensive documentation on how it works and there is a vignette here:

https://bioconductor.org/packages/release/bioc/vignettes/batchelor/inst/doc/correction.html

The original paper is here:

https://pubmed.ncbi.nlm.nih.gov/29608177/

Everything we need is included in the `Batchelor` library available in bioconductor. `Batchelor` is from the same authors as `scater` and `scran` and also features in the `OSCA` manual.

### Check presence of batch effect

Batchelor commands to make the two batches and identify highly variable genes for faster dimensionality reduction. It has a slightly different workflow so here we demonstrate from the beginning again. We remake separate sce objects from our combined sample.

```{r noCor_prep_batch_GSM3872442}

library(batchelor)

sce1 <- sample3.sce[, sample3.sce$Run == "SRR9264351"]
sce2 <- sample3.sce[, sample3.sce$Run == "SRR9264352"]
```

```{r noCor_var_batch_GSM3872442}
library(scran)
dec1 <- modelGeneVar(sce1)
dec2 <- modelGeneVar(sce2)
combined.dec <- combineVar(dec1, dec2)
chosen.hvgs <- combined.dec$bio > 0
summary(chosen.hvgs)

```

As a diagnostic, we check that there actually is a batch effect across these datasets by checking that they cluster separately. Here, we combine the two SingleCellExperiment objects without any correction using the NoCorrectParam() flag, and we informally verify that cells from different batches are separated using a t-SNE plot.

There is a moderate batch effect.

```{r noCor_redDim_batch_GSM3872442}
library(scater)
combined <- correctExperiments(A=sce1, B=sce2, PARAM=NoCorrectParam())
combined <- runPCA(combined, subset_row=chosen.hvgs)
combined <- runTSNE(combined, dimred="PCA")
combined <- runUMAP(combined, dimred="PCA")
plotPCA(combined, colour_by="batch")
plotTSNE(combined, colour_by="batch")
plotUMAP(combined, colour_by="batch")
```

```{r noCor_addNoCor_batch_GSM3872442}
reducedDim(sample3.sce, "PCA_noCor") <- reducedDim(combined, "PCA")
reducedDim(sample3.sce, "TSNE_noCor") <- reducedDim(combined, "TSNE")
reducedDim(sample3.sce, "UMAP_noCor") <- reducedDim(combined, "UMAP")

```

### Correct batch effect with mnnCorrect

Gene expression values are used to identify cells with similar expression patterns in both batches.

We need the normalised counts:

```{r mnnCor_prep_batch_GSM3872442}
batch1 <- logcounts(sce1)
batch2 <- logcounts(sce2)

```
Now we can run mnnCorrect, even with 2 samples it can take a while.

```{r mnnCor_hvgOnly_batch_GSM3872442}
# returns a matrix with rownames only for the gene subset,
# at the top of the matrix
# preventing copy of that corrected matrix as an assay in the SCE object

# mmnCorrect returns the corrected gene expression matrix directly

y <- batchelor::mnnCorrect(
          batch1, batch2,  
	  correct.all = TRUE,
          k = 20,
          sigma = 0.1,
          cos.norm.in = TRUE,
          svd.dim = 2
        )
```

Copy the corrected values to the SCE object:

```{r mnnCor_addCorr_batch_GSM3872442}
assay(sample3.sce, "mnn") <- assay(y, "corrected")
```

Show reduced dimension plots and check for improved mixing of cells from the two sets:

```{r mnnCor_pca_batch_GSM3872442}
sample3.sce <- runPCA(sample3.sce, exprs_values = "mnn")
plotPCA(sample3.sce, colour_by="batch")
reducedDim(sample3.sce, "PCA_mnn") <- reducedDim(sample3.sce, "PCA")
```

```{r mnnCor_tsne_batch_GSM3872442}
sample3.sce <- runTSNE(sample3.sce, dimred="PCA_mnn")
plotTSNE(sample3.sce, colour_by="batch")
reducedDim(sample3.sce, "TSNE_mnn") <- reducedDim(sample3.sce, "TSNE")
```

```{r mnnCor_umap_batch_GSM3872442}
sample3.sce <- runUMAP(sample3.sce, dimred="PCA_mnn")
plotUMAP(sample3.sce, colour_by="batch")
reducedDim(sample3.sce, "UMAP_mnn") <- reducedDim(sample3.sce, "UMAP")
```

The plots show nice amounts of mixing.

## fastMNN

This method is faster than mnnCorrect as it identifies nearest neighbours after dimensionality reduction and is also avaliable in the `Batchelor` package. 

```{r fastMnn_run_batch_GSM3872442}
fx <- batchelor::fastMNN(
                      sample3.sce,
                      batch = sample3.sce$Run
			)

```

Copy the corrected values to the SCE object:

```{r fastMnn_addRecon_batch_GSM3872442}
# fastMNN may drop some genes
# so we may not be able to keep the outcome in 'assay'
assay(sample3.sce, "fastmnn") <- assay(fx, "reconstructed")
```

Show reduced dimension plots and check for improved mixing of cells from the two sets:

```{r fastMnn_pca_batch_GSM3872442}
fastmnn_pca <- runPCA(assay(sample3.sce, "fastmnn"), rank=2) # slow
reducedDim(sample3.sce, "PCA_fastmnn") <- fastmnn_pca$rotation

plotReducedDim(
  sample3.sce,
  dimred = "PCA_fastmnn",
  colour_by = "batch",
  size_by = "sum",
  shape_by = "Sample.Name"
) + ggtitle("PCA plot: fastMNN") 
```

```{r fastMnn_tsne_batch_GSM3872442}
sample3.sce <- runTSNE(sample3.sce, dimred="PCA_fastmnn")
plotTSNE(sample3.sce, colour_by="batch")
reducedDim(sample3.sce, "TSNE_fastmnn") <- reducedDim(sample3.sce, "TSNE")
```

```{r fastMnn_umap_batch_GSM3872442}
sample3.sce <- runUMAP(sample3.sce, dimred="PCA_fastmnn")
plotUMAP(sample3.sce, colour_by="batch")
reducedDim(sample3.sce, "UMAP_fastmnn") <- reducedDim(sample3.sce, "UMAP")
```

## Harmony

Harmony [Korsunsky2018fast] is a newer batch correction method, which is also designed to operate on PC space. The algorithm proceeds to iteratively cluster the cells, with the objective function formulated to promote cells from multiple datasets within each cluster. Once a clustering is obtained, the positions of the centroids of each dataset are obtained on a per-cluster basis and the coordinates are corrected. This procedure is iterated until convergence. Harmony comes with a theta parameter that controls the degree of batch correction (higher values lead to more dataset integration), and can account for multiple experimental and biological factors on input (see [variant of the 'Hemberg course'](https://biocellgen-public.svi.edu.au/mig_2019_scrnaseq-workshop/public/normalization-confounders-and-batch-correction.html#harmony)).

```{r harmony_run_batch_GSM3872442}
library(harmony)

reducedDim(sample3.sce, "PCA_logcounts") <- reducedDim(
  runPCA(sample3.sce, exprs_values = "logcounts")
)

#Seeing how the end result of Harmony is an altered dimensional reduction space created on the basis of PCA, we plot the obtained manifold here and exclude it from the rest of the follow-ups in the section.

pca <- as.matrix(reducedDim(sample3.sce, "PCA_logcounts"))
harmony_emb <- HarmonyMatrix(pca,
			     sample3.sce$batch,
			     theta=2,
			     do_pca=FALSE)
reducedDim(sample3.sce, "harmony") <- harmony_emb

plotReducedDim(
    sample3.sce,
    dimred = 'harmony',
    colour_by = "batch",
    size_by = "sum",
    shape_by = "Sample.Name"
)
```

## Session information

<details>
```{r}
sessionInfo()
```
</details>
