---
title: "CRUK CI Summer School 2021 - Introduction to single-cell RNA-seq analysis"
subtitle: 'Cluster marker genes - Exercise'
author: "Stephane Ballereau, Zeynep Kalender Atak"
output:
  html_document:
    toc: yes
---


# Introduction

In the course materials we focused on differential expression analysis of cluster 8. For this exercise we would like you to perform similar analysis for cluster 6. 

## Load packages

```{r libraries_preProc, warning=FALSE, message=FALSE}
library(ggplot2)
library(scater)
library(scran)
library(dplyr)
library(RColorBrewer)
library(pheatmap)
library(glue)
```

##  Load the data

```{r}
# load both sce objects
uncorrected <- readRDS("~/Course_Materials/scRNAseq/Robjects/DataIntegration_uncorrected.rds")
corrected <- readRDS("~/Course_Materials/scRNAseq/Robjects/caron_postDeconv_5hCellPerSpl_dsi_PBMMC_ETV6-RUNX1_clust.Rds")
``` 

## 1. Perform differential expression analysis. 

Extract louvain clusters from the corrected SCE object, and use the uncorrected SCE objects with `findMarkers` command. You can select any of the testing method and options. 

```{r}
clusters.mnn <- factor(paste0("c",corrected$louvain))

markers.out <- findMarkers(uncorrected, 
                           groups=clusters.mnn, 
                           direction="up", 
                           block=uncorrected$SampleGroup,
                           row.data=rowData(uncorrected))
```

## 2. Obtain markers for cluster 11 

```{r}
c11_markers <- markers.out[["c11"]]
head(c11_markers)
```

### 2a. Visualize one of the marker genes using violin plot 

Here you need to use `uncorrected` SCE object (and Ensembl gene IDs)

```{r}
plotExpression(uncorrected,
               x=I(factor(corrected$louvain)),
               features="ENSG00000019582", # "CD74",
               colour_by="SampleGroup") +
  facet_wrap(~colour_by) 
```

## 2b. Visualize one of the markers on tSNE plot 

Here you can use `corrected` SCE object (use gene symbols). 

```{r}
plotTSNE(corrected, colour_by="CD74", by_exprs_values = "reconstructed")
```

## 3. Take top 5 genes from each pairwise comparison and create a heatmap (hint: use the `Top` field)

```{r}
c11_top5 <- c11_markers[c11_markers$Top <= 5,]
c11_top5_logFC <- getMarkerEffects(c11_top5)

# change rownames from Ensembl gene IDs to gene symbols 
rownames(c11_top5_logFC) <- rowData(uncorrected)[rownames(c11_top5_logFC), "Symbol"]

pheatmap(c11_top5_logFC, breaks=seq(-5, 5, length.out=101))

```

